File: modules/common/src/main/java/com/sygic/modules/common/di/module/AppModule.java
package com.sygic.modules.common.di.module;

import android.app.Application;
import androidx.fragment.app.Fragment;
import dagger.Module;
import dagger.Provides;

import javax.inject.Singleton;

@Module
public class AppModule {

    private final Application app;

    public AppModule(final Fragment fragment) {
        app = fragment.requireActivity().getApplication();
    }

    @Singleton
    @Provides
    Application provideApplication() {
        return app;
    }
}

File: app/samples/src/main/java/com/sygic/samples/di/AppModule.kt
package com.sygic.samples.di

import android.app.Application
import android.content.Context
import android.content.SharedPreferences
import com.sygic.samples.app.App
import com.sygic.samples.demo.PersistentRoutingOptionsManager
import com.sygic.samples.demo.RoutingOptionsManager
import dagger.Binds
import dagger.Module
import dagger.Provides
import javax.inject.Named
import javax.inject.Singleton

@Module
internal abstract class AppModule {
    @Module
    companion object {
        private const val ROUTING_OPTIONS_PREFERENCES = "routing_options_preferences"

        @JvmStatic
        @Singleton
        @Provides
        @Named(ROUTING_OPTIONS_PREFERENCES)
        fun provideRoutingOptionsPreferences(app: App): SharedPreferences
                = app.getSharedPreferences(ROUTING_OPTIONS_PREFERENCES, Context.MODE_PRIVATE)

        @JvmStatic
        @Singleton
        @Provides
        fun provideRoutingOptionsManager(
            @Named(ROUTING_OPTIONS_PREFERENCES) preferences: SharedPreferences
        ): RoutingOptionsManager = PersistentRoutingOptionsManager(preferences)
    }

    @Binds
    abstract fun bindApplication(app: App): Application
}